Automattic/woocommerce-payments
# Add header to Payment Methods, tweak display when UPE is disabled.

Fixes #2175 

#### Changes proposed in this Pull Request

Add a header to the Payment Methods Card component which will allow(when UPE is enabled) users to submit feedback and disable UPE.

There are essentially 4 state combinations now:

- UPE Settings Preview **Enabled** / UPE **Enabled** - User sees Payment Methods component and header component(Feedback/Disable).
- UPE Settings Preview **Enabled** / UPE **Disabled** - User sees Payment Methods component, no header component, and sees a button to enable UPE.
- UPE Settings Preview **Disabled** / UPE **Enabled** - This _shouldn't_ be possible without manually enabling UPE in the DB. Nothing breaks but perhaps we should consider logic routes if we end up _reverting_ the UPE Settings Preview for a user(disabling after previously enabled).
- UPE Settings Preview **Disabled** / UPE **Disabled** - User sees no Payment Methods area at all.

In action:
![Screen Capture on 2021-07-06 at 14-18-34](https://user-images.githubusercontent.com/6723003/124732542-93aea100-df13-11eb-8a9c-27cc6e00b52c.gif)


#### Testing instructions

To test enabling/disabling UPE Settings Preview, check/uncheck `Enable UPE settings changes` on the WCPay Dev page.

To test enabling/disabling UPE itself being availabile, check/uncheck `Enable UPE checkout` on the WCPay Dev page.

Some notes:

I wrote three new tests but I'm brand new to the React Testing Library so let me know if you see something that could be improved or if there are cases I'm missing.

In the `SettingsManager` component I am grabbing the "UPE enabled" featureFlag and passing it down to the `PaymentMethods` component, for two reasons:

1. Easier to test through injection of `isUPEEnabled` as a prop rather than mocking state within the PaymentMethods component.
2. There may be other components within the SettingsManager component that would care about whether or not UPE is enabled so it might be a good thing to have it available at that level.

I bring this up because technically only the `PaymentMethods` component needs `isUPEEnabled` at this time which might tempt one to keep it contained within the `PaymentMethods` component. I'd like to hear any counterarguments to my decision to _not_ do that if there are some. :)

- Because we need to be able to enable/disable UPE locally in this component, I grabbed the passed-in value for `isUPEEnabled` and [initialized state to that value](https://github.com/Automattic/woocommerce-payments/pull/2372/files#diff-cce9c1ee767cec1c20dc9ee0ce12f48dba6e2b2f9aaf816bf06eec75ca076b96R98) so that it can be updated with useState(). This feels awkward but _should_ be removed when we have real actions we can dispatch to update that feature flag. I'm curious if that's a terrible anti-pattern or not. 🤷 

-------------------
